File: Fledge/FledgeJava/app/src/main/java/com/example/adservices/samples/fledge/WaterfallMediationHelpers/CustomAudienceHelper.java
package com.example.adservices.samples.fledge.WaterfallMediationHelpers;

import static com.example.adservices.samples.fledge.WaterfallMediationHelpers.Constants.BIDDING_LOGIC_JS;
import static com.example.adservices.samples.fledge.WaterfallMediationHelpers.Constants.BIDDING_URI_SUFFIX;
import static com.example.adservices.samples.fledge.WaterfallMediationHelpers.Constants.BID_SIGNALS_FORMAT;
import static com.example.adservices.samples.fledge.WaterfallMediationHelpers.Constants.DEFAULT_BASE_URI_FORMAT;
import static com.example.adservices.samples.fledge.WaterfallMediationHelpers.Constants.TAG;
import static com.example.adservices.samples.fledge.WaterfallMediationHelpers.Constants.TRUSTED_BIDDING_URI_SUFFIX;

import android.adservices.customaudience.AddCustomAudienceOverrideRequest;
import android.net.Uri;
import android.util.Log;

import androidx.annotation.RequiresApi;
import androidx.privacysandbox.ads.adservices.common.AdData;
import androidx.privacysandbox.ads.adservices.common.AdSelectionSignals;
import androidx.privacysandbox.ads.adservices.common.AdTechIdentifier;
import androidx.privacysandbox.ads.adservices.customaudience.CustomAudience;
import androidx.privacysandbox.ads.adservices.customaudience.TrustedBiddingData;

import com.example.adservices.samples.fledge.clients.CustomAudienceClient;
import com.example.adservices.samples.fledge.clients.TestCustomAudienceClient;
import com.example.adservices.samples.fledge.sampleapp.MainActivity;

import org.json.JSONObject;

import java.time.Duration;
import java.time.Instant;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

/** Helps creating CAs to set the environment for mediation activity. */
@RequiresApi(api = 34)
public class CustomAudienceHelper {

    private final CustomAudienceClient customAudienceClient;
    private final TestCustomAudienceClient testCustomAudienceClient;

    public CustomAudienceHelper(
            CustomAudienceClient customAudienceClient,
            TestCustomAudienceClient testCustomAudienceClient) {
        this.customAudienceClient = customAudienceClient;
        this.testCustomAudienceClient = testCustomAudienceClient;
    }

    public AdTechIdentifier configureCustomAudience(
            String customAudienceName, double bid, Uri baseUri, boolean useOverrides) {
        String uriFriendlyName = Constants.uriFriendlyString(customAudienceName);
        if (useOverrides) {
            baseUri = Uri.parse(String.format(DEFAULT_BASE_URI_FORMAT, uriFriendlyName));
        }
        CustomAudience customAudience = getCustomAudience(customAudienceName, baseUri, bid);

        if (useOverrides) {
            addOverrideCustomAudience(customAudience);
        }

        joinCustomAudience(customAudience);
        Log.i(TAG, customAudience.getBuyer() + " buyer is returned");
        return customAudience.getBuyer();
    }

    private void addOverrideCustomAudience(CustomAudience customAudience) {

        String biddingLogicJs =
                String.format(
                        BIDDING_LOGIC_JS, Constants.uriFriendlyString(customAudience.getName()));
        try {
            testCustomAudienceClient
                    .overrideCustomAudienceRemoteInfo(
                            new AddCustomAudienceOverrideRequest.Builder()
                                    .setBuyer(
                                            android.adservices.common.AdTechIdentifier.fromString(
                                                    customAudience.getBuyer().getIdentifier()))
                                    .setName(customAudience.getName())
                                    .setBiddingLogicJs(biddingLogicJs)
                                    .setTrustedBiddingSignals(
                                            android.adservices.common.AdSelectionSignals.fromString(
                                                    "{}"))
                                    .build())
                    .get(10, TimeUnit.SECONDS);
        } catch (Exception e) {
            Log.e(MainActivity.TAG, "Exception calling overrideCustomAudienceRemoteInfo", e);
        }
    }

    private void joinCustomAudience(CustomAudience customAudience) {
        try {
            customAudienceClient.joinCustomAudience(customAudience).get(10, TimeUnit.SECONDS);
            Thread.sleep(1000);
        } catch (Exception e) {
            Log.e(MainActivity.TAG, "Exception calling joinCustomAudience", e);
        }
    }

    private CustomAudience getCustomAudience(String customAudienceName, Uri baseUri, double bid) {
        String uriFriendlyName = Constants.uriFriendlyString(customAudienceName);

        AdSelectionSignals userBiddingSignals =
                new AdSelectionSignals(String.format(BID_SIGNALS_FORMAT, bid));
        AdTechIdentifier buyer = new AdTechIdentifier(baseUri.getHost());
        Instant activationTime = Instant.now();
        Instant expirationTime = Instant.now().plus(Duration.ofDays(1));
        List<AdData> ads =
                Collections.singletonList(
                        new AdData(
                                getAdRenderUri(baseUri, uriFriendlyName),
                                new JSONObject().toString()));
        TrustedBiddingData trustedBiddingData =
                new TrustedBiddingData(getTrustedBiddingDataUri(baseUri), Collections.emptyList());

        return new CustomAudience(
                buyer,
                customAudienceName,
                getDailyUpdateUri(baseUri),
                getBiddingLogicUri(baseUri),
                ads,
                activationTime,
                expirationTime,
                userBiddingSignals,
                trustedBiddingData);
    }

    private Uri getBiddingLogicUri(Uri baseUri) {
        return baseUri.buildUpon().appendPath(BIDDING_URI_SUFFIX).build();
    }

    private Uri getDailyUpdateUri(Uri baseUri) {
        return baseUri.buildUpon().appendPath(BIDDING_URI_SUFFIX).build();
    }

    private Uri getTrustedBiddingDataUri(Uri baseUri) {
        return baseUri.buildUpon().appendPath(TRUSTED_BIDDING_URI_SUFFIX).build();
    }

    private Uri getAdRenderUri(Uri baseUri, String adName) {
        return baseUri.buildUpon().appendPath(BIDDING_URI_SUFFIX).appendPath(adName).build();
    }
}

File: Fledge/FledgeKotlin/app/src/main/java/com/example/adservices/samples/fledge/waterfallmediationhelpers/CustomAudienceHelper.kt
package com.example.adservices.samples.fledge.waterfallmediationhelpers

import androidx.annotation.RequiresApi
import com.example.adservices.samples.fledge.clients.CustomAudienceClient
import com.example.adservices.samples.fledge.clients.TestCustomAudienceClient
import android.adservices.common.AdTechIdentifier
import android.adservices.customaudience.CustomAudience
import android.adservices.customaudience.AddCustomAudienceOverrideRequest
import android.adservices.common.AdSelectionSignals
import com.example.adservices.samples.fledge.sampleapp.MainActivity
import android.adservices.common.AdData
import org.json.JSONObject
import android.adservices.customaudience.TrustedBiddingData
import android.net.Uri
import android.util.Log
import com.example.adservices.samples.fledge.waterfallmediationhelpers.Constants.TAG
import java.lang.Exception
import java.time.Duration
import java.time.Instant
import java.util.concurrent.TimeUnit

/**
 * Helps creating CAs to set the environment for mediation activity.
 */
@RequiresApi(api = 34)
class CustomAudienceHelper(
  private val customAudienceClient: CustomAudienceClient,
  private val testCustomAudienceClient: TestCustomAudienceClient
) {
  fun configureCustomAudience(
    customAudienceName: String,
    bid: Double,
    baseUri: Uri,
    useOverrides: Boolean
  ): AdTechIdentifier {
    var baseUri = baseUri
    val uriFriendlyName = Constants.uriFriendlyString(customAudienceName)
    if (useOverrides) {
      baseUri = Uri.parse(String.format(Constants.DEFAULT_BASE_URI_FORMAT, uriFriendlyName))
    }
    val customAudience = getCustomAudience(customAudienceName, baseUri, bid)
    if (useOverrides) {
      addOverrideCustomAudience(customAudience)
    }
    joinCustomAudience(customAudience)
    Log.i(Constants.TAG, customAudience.buyer.toString() + " buyer is returned")
    return customAudience.buyer
  }

  private fun addOverrideCustomAudience(customAudience: CustomAudience) {
    val biddingLogicJs =
      String.format(Constants.BIDDING_LOGIC_JS, Constants.uriFriendlyString(customAudience.name))
    try {
      testCustomAudienceClient.overrideCustomAudienceRemoteInfo(
        AddCustomAudienceOverrideRequest.Builder()
          .setBuyer(customAudience.buyer)
          .setName(customAudience.name)
          .setBiddingLogicJs(biddingLogicJs)
          .setTrustedBiddingSignals(AdSelectionSignals.EMPTY)
          .build())!![10, TimeUnit.SECONDS]
    } catch (e: Exception) {
      Log.e(TAG, "Exception calling overrideCustomAudienceRemoteInfo", e)
    }
  }

  private fun joinCustomAudience(customAudience: CustomAudience) {
    try {
      customAudienceClient.joinCustomAudience(customAudience)[10, TimeUnit.SECONDS]
      Thread.sleep(1000)
    } catch (e: Exception) {
      Log.e(TAG, "Exception calling joinCustomAudience", e)
    }
  }

  private fun getCustomAudience(
    customAudienceName: String,
    baseUri: Uri,
    bid: Double
  ): CustomAudience {
    val uriFriendlyName = Constants.uriFriendlyString(customAudienceName)
    val userBiddingSignals =
      AdSelectionSignals.fromString(String.format(Constants.BID_SIGNALS_FORMAT, bid))
    val buyer = AdTechIdentifier.fromString(baseUri.host!!)
    val activationTime = Instant.now()
    val expirationTime = Instant.now().plus(Duration.ofDays(1))
    val ads = listOf(AdData.Builder()
                       .setRenderUri(getAdRenderUri(baseUri, uriFriendlyName))
                       .setMetadata(JSONObject().toString())
                       .build())
    val trustedBiddingData = TrustedBiddingData.Builder()
      .setTrustedBiddingKeys(emptyList())
      .setTrustedBiddingUri(getTrustedBiddingDataUri(baseUri))
      .build()
    return CustomAudience.Builder()
      .setBuyer(buyer)
      .setName(customAudienceName)
      .setActivationTime(activationTime)
      .setExpirationTime(expirationTime)
      .setDailyUpdateUri(getDailyUpdateUri(baseUri))
      .setUserBiddingSignals(userBiddingSignals)
      .setTrustedBiddingData(trustedBiddingData)
      .setBiddingLogicUri(getBiddingLogicUri(baseUri))
      .setAds(ads)
      .build()
  }

  private fun getBiddingLogicUri(baseUri: Uri): Uri {
    return baseUri.buildUpon().appendPath(Constants.BIDDING_URI_SUFFIX).build()
  }

  private fun getDailyUpdateUri(baseUri: Uri): Uri {
    return baseUri.buildUpon().appendPath(Constants.BIDDING_URI_SUFFIX).build()
  }

  private fun getTrustedBiddingDataUri(baseUri: Uri): Uri {
    return baseUri.buildUpon().appendPath(Constants.TRUSTED_BIDDING_URI_SUFFIX).build()
  }

  private fun getAdRenderUri(baseUri: Uri, adName: String?): Uri {
    return baseUri.buildUpon().appendPath(Constants.BIDDING_URI_SUFFIX).appendPath(adName).build()
  }
}